### PR TITLE
Stop requiring imported pubkey to sign non-PKH schemes

### PR DESCRIPTION
**Pull request by sipa**
...

This is a fix for #14415

### CODE REVIEW

**Comment by promag** [comment]
Test failing https://github.com/bitcoin/bitcoin/blob/a23e50fcd175a038690bd196cf0573fd98435e47/src/wallet/test/psbt_wallet_tests.cpp#L71

**Comment by MeshCollider** [comment]
This is a fix for #14415 right?

**Comment by sipa** [comment]
@MeshCollider Hopefully.

File: src/script/sign.cpp [hunk 1]
```diff
@@ -73,15 +73,14 @@ static bool GetPubKey(const SigningProvider& provider, SignatureData& sigdata, c
...
+static bool CreateSig(const BaseSignatureCreator& creator, SignatureData& sigdata, const SigningProvider& provider, std::vector<unsigned char>& sig_out, const CPubKey& pubkey, const CScript& scriptcode, SigVersion sigversion)
 {
+    CKeyID keyid = pubkey.GetID();
     const auto it = sigdata.signatures.find(keyid);
     if (it != sigdata.signatures.end()) {
         sig_out = it->second.second;
         return true;
     }
-    CPubKey pubkey;
-    GetPubKey(provider, sigdata, keyid, pubkey);
```
Comment (achow101) [inline src/script/sign.cpp]: Removing this is also breaking the PSBT updating test. Without it, we don't know which pubkeys are needed for signing in order to properly update a PSBT with the pubkeys.

Comment (instagibbs) [reply src/script/sign.cpp]: sorry can you point to where this breaks stuff exactly, non-obvious to me

Comment (achow101) [reply src/script/sign.cpp]: PSBT relies on the `GetPubKey` helper function to find out which public keys are used and to add the public key to a `SignatureData` object. The public keys are taken from the sigdata after `ProduceSignature` is done and moved into the PSBT. Without this call to `GetPubKey`, the pubkeys that were needed during signing are not added to the sigdata and thus not added to the PSBT.

Comment (sipa) [reply src/script/sign.cpp]: Possibly fixed.

**Comment by instagibbs** [comment]
one build was failing to finish building(?), kicked

**Comment by achow101** [comment]
utACK 2f6b466aeb6d4c88ab2e0e8b2a402be0743608b5

**Comment by instagibbs** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/14424/commits/2f6b466aeb6d4c88ab2e0e8b2a402be0743608b5

**Comment by sipa** [comment]
Marked as 0.17.1.

**Comment by MeshCollider** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/14424/commits/2f6b466aeb6d4c88ab2e0e8b2a402be0743608b5

**Comment by jonasschnelli** [comment]
Tested ACK 2f6b466aeb6d4c88ab2e0e8b2a402be0743608b5, fixes #14415

**Comment by fanquake** [comment]
Backported in #14889.

**Pull request merged**